synapse/handlers/sliding_sync.py

```diff
@@ -619,6 +619,31 @@ async def current_sync_for_user(
         # Fetch room data
         rooms: Dict[str, SlidingSyncResult.RoomResult] = {}
 
+        # Filter out rooms that haven't received updates and we've sent down
+        # previously.
+        if from_token:
+            rooms_should_send = set()
+            for room_id in relevant_room_map:
+                status = await self.connection_store.have_sent_room(
+                    sync_config,
+                    from_token.connection_position,
+                    room_id,
+                )
+                if status.status != HaveSentRoomFlag.LIVE:
+                    rooms_should_send.add(room_id)
+
+            # We only need to check for any new events and not state changes, as
+            # state changes can only happen if an event has also been sent.
+            rooms_that_have_updates = self.store.get_rooms_that_might_have_updates(
+                relevant_room_map.keys(), from_token.stream_token.room_key
```

Comment:
We could make the optimization here to only check rooms not already added to `rooms_should_send`:

```suggestion
                set(relevant_room_map.keys()).difference(rooms_should_send), from_token.stream_token.room_key
```

It looks like `get_rooms_that_might_have_updates(...)` is already checking an in-memory cache so maybe it's not worth it.

Reply:
Yeah, I think constructing a new set is probably going to be more effort than this removes :shrug: